### PR TITLE
Added Content for Build Systems

### PR DESCRIPTION
**Pull request by SuperChamp234**
- Added content for Build Systems (#58) . We explain the compilation process, the need for build systems, and introduce the GNU Make Build System.
- The GNU Make build system is explained from a basic level. We cover:
         - The basic Makefile syntax.
         - Variables in Makefile.
         - Automatic Variables in Makefile.
         - Wildcards in Makefile.
         - Commands in Makefile.
         - Three example exercises for easing the readers into the creation of Makefile.

Co-authored by @PritK99.

### CODE REVIEW

File: 2_build_systems/2_build_systems.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+For example, the ```#include``` directive incorporates the contents of a header file to the compilation. Similarly, The ```#define``` directive is used to define a macro.
+
+### Different types of files during compilation
+
+<p align="center">
+  <img src="../../assets/images/Compilation_Stages.png" width="800" />
+</p>
+
+The compilation process transforms a human-readable code into a machine-readable format. For a programming language, it happens before a program starts executing to check the syntax and semantics of the code. The compilation process involves four steps: pre-processing, compiling, assembling, and linking then, we run the obtained executable file to get an output on the screen.
```
Comment (amanchhaparia) [inline 2_build_systems/2_build_systems.md:19]: Can add the detailed explanation for each compilation process steps.

File: 2_build_systems/2_build_systems.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+### Compiler Directives 
+
+Compiler directives refer to statement written in the source code of a program that lets the programmer instruct the compiler to perform a specific operation within the compilation phase itself.
+
+For example, the ```#include``` directive incorporates the contents of a header file to the compilation. Similarly, The ```#define``` directive is used to define a macro.
+
+### Different types of files during compilation
+
+<p align="center">
+  <img src="../../assets/images/Compilation_Stages.png" width="800" />
```
Comment (amanchhaparia) [inline 2_build_systems/2_build_systems.md:16]: Incorrect image path

File: 2_build_systems/2_build_systems.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+The compilation process transforms a human-readable code into a machine-readable format. For a programming language, it happens before a program starts executing to check the syntax and semantics of the code. The compilation process involves four steps: pre-processing, compiling, assembling, and linking then, we run the obtained executable file to get an output on the screen.
+
+* Source(.c):
+These files contain function definitions, and the entire program logics, these files are human readable and by convention their names ending with . c
+
+* Assembly(.s):
+Assembly Language mainly consists of mnemonic processor instructions or data and other statements or instructions.Flag -S eg. gcc -S foo.c
+
+* Object(.o):
+These files are produced as the output of the compiler. They consist of function definitions in binary form, but they are not executable by themselves and by convention their names end with .o.Flag -o eg. gcc -o foo.c
```
Comment (amanchhaparia) [inline 2_build_systems/2_build_systems.md:28]: image
Can improve the readability using the proper formatting style like: 

These files are produced as the output of the compiler. They consist of function definitions in binary form, but they are not executable by themselves and by convention their names end with `.o`. 
`Flag: -o` 
`eg`:  gcc -o foo.c

File: 2_build_systems/2_build_systems.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+```c
+gcc main.c
+```
+This instructs the compiler to take convert main.c into a binary class file. In the simplest case, this is we you need.
+
+However, as soon as code expands, the complications begin. Our compiler has no way of finding code stored in other parts of the filesystem (perhaps a library shared by several projects). It also only knows how to build the code. 
+
+Large systems often involve different pieces written in a variety of programming languages with webs of dependencies among those pieces, meaning no compiler for a single language can possibly build the entire system. Thus, there are multiple dependencies and without a build system, one must evaluate what the code depends on and build those pieces in the proper orde
+The compiler also doesn’t know anything about how to handle external dependencies. Without a build system, you could manage this by downloading the dependency from the internet, sticking it in a lib folder on the hard drive, and configuring the compiler to read libraries from that directory. But over time, it's difficult to maintain the updates, versions, and source of these external dependencies.
+
```
Comment (amanchhaparia) [inline 2_build_systems/2_build_systems.md:51]: Too much theory, try to improve it.

File: 2_build_systems/2_build_systems.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+The need for a build system might not be immediately obvious We can invoke tools like gcc directly from the command line, or the equivalent in an integrated development environment (IDE). As long as all the source code is in the same directory, a command like this works fine:
+
+```c
+gcc main.c
+```
+This instructs the compiler to take convert main.c into a binary class file. In the simplest case, this is we you need.
+
+However, as soon as code expands, the complications begin. Our compiler has no way of finding code stored in other parts of the filesystem (perhaps a library shared by several projects). It also only knows how to build the code. 
+
+Large systems often involve different pieces written in a variety of programming languages with webs of dependencies among those pieces, meaning no compiler for a single language can possibly build the entire system. Thus, there are multiple dependencies and without a build system, one must evaluate what the code depends on and build those pieces in the proper orde
```
Comment (amanchhaparia) [inline 2_build_systems/2_build_systems.md:49]: typo error `orde`

**Review by amanchhaparia — CHANGES_REQUESTED** [review]

**Comment by PritK99** [comment]
Made the following changes :- 

1) Added detailed explanation for all compilation stages
2) Resolved the path of image : Compilation_Stages.png
3) Used proper formatting to increase readability
4) Made the theory to the point by removing the redundant theory
5) Resolved the typo error

Please review.

**Comment by amanchhaparia** [comment]
@PritK99, kindly mark as `resolve conversation` once the required changes had been made.

File: 2_build_systems/2_build_systems.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+3) Assembling : The assembly code is converted into object code by using an assembler. The name of the object file generated by the assembler is the same as the source file. 
+
+4) Linking : The main working of the linker is to combine the object code of library files with the object code of our program.
+
+### Different types of files during compilation
+
+During the stages of compilation, there are several types of files involved.
+
+* Source(.c):
+These files contain function definitions, and the entire program logics, these files are human readable and by convention their names ending with ```. c```
```
Comment (amanchhaparia) [inline 2_build_systems/2_build_systems.md:34]: typo error, should be `.c`

**Review by amanchhaparia — CHANGES_REQUESTED** [review]

File: 2_build_systems/2_build_systems.md (hunk 2 above)
Comment (SuperChamp234) [reply 2_build_systems/2_build_systems.md]: Resolved

File: 2_build_systems/2_build_systems.md (hunk 1 above)
Comment (SuperChamp234) [reply 2_build_systems/2_build_systems.md:19]: Resolved in latest commit

File: 2_build_systems/2_build_systems.md (hunk 3 above)
Comment (SuperChamp234) [reply 2_build_systems/2_build_systems.md]: Resolved in latest commit

File: 2_build_systems/2_build_systems.md (hunk 5 above)
Comment (SuperChamp234) [reply 2_build_systems/2_build_systems.md]: Resolved in latest commit

File: 2_build_systems/2_build_systems.md (hunk 6 above)
Comment (SuperChamp234) [reply 2_build_systems/2_build_systems.md]: Resolved in latest commit

**Comment by PritK99** [comment]
Made a few corrections in 2_build_systems.md in the latest commit. Please review.

**Pull request merged**